Flaky-test: MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting

MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting is flaky. The test fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/8706/checks?check_run_id=1457916704#step:9:884)

```
[INFO] Running org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest
Error:  Tests run: 6, Failures: 3, Errors: 0, Skipped: 3, Time elapsed: 8.06 s <<< FAILURE! - in org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest
Error:  (org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)  Time elapsed: 0.016 s  <<< FAILURE!
java.lang.RuntimeException: broker already started!
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.startBroker(MockedPulsarServiceBaseTest.java:245)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.init(MockedPulsarServiceBaseTest.java:157)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.internalSetup(MockedPulsarServiceBaseTest.java:102)
	at org.apache.pulsar.broker.service.BrokerTestBase.baseSetup(BrokerTestBase.java:41)
	at org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting(MessagePublishBufferThrottleTest.java:117)
```

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
[INFO] Running org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest
Error:  Tests run: 6, Failures: 3, Errors: 0, Skipped: 3, Time elapsed: 8.06 s <<< FAILURE! - in org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest
Error:  (org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)  Time elapsed: 0.016 s  <<< FAILURE!
java.lang.RuntimeException: broker already started!
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.startBroker(MockedPulsarServiceBaseTest.java:245)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.init(MockedPulsarServiceBaseTest.java:157)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.internalSetup(MockedPulsarServiceBaseTest.java:102)
	at org.apache.pulsar.broker.service.BrokerTestBase.baseSetup(BrokerTestBase.java:41)
	at org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting(MessagePublishBufferThrottleTest.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Error:  testMessagePublishBufferThrottleDisabled(org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)  Time elapsed: 0.013 s  <<< FAILURE!
java.lang.RuntimeException: broker already started!
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.startBroker(MockedPulsarServiceBaseTest.java:245)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.init(MockedPulsarServiceBaseTest.java:157)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.internalSetup(MockedPulsarServiceBaseTest.java:102)
	at org.apache.pulsar.broker.service.BrokerTestBase.baseSetup(BrokerTestBase.java:41)
	at org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testMessagePublishBufferThrottleDisabled(MessagePublishBufferThrottleTest.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Error:  testMessagePublishBufferThrottleEnable(org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)  Time elapsed: 0.012 s  <<< FAILURE!
java.lang.RuntimeException: broker already started!
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.startBroker(MockedPulsarServiceBaseTest.java:245)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.init(MockedPulsarServiceBaseTest.java:157)
	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.internalSetup(MockedPulsarServiceBaseTest.java:102)
	at org.apache.pulsar.broker.service.BrokerTestBase.baseSetup(BrokerTestBase.java:41)
	at org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testMessagePublishBufferThrottleEnable(MessagePublishBufferThrottleTest.java:80)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:  org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting(org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)
[INFO]   Run 1: PASS
Error:    Run 2: MessagePublishBufferThrottleTest.testBlockByPublishRateLimiting:117->BrokerTestBase.baseSetup:41->MockedPulsarServiceBaseTest.internalSetup:102->MockedPulsarServiceBaseTest.init:157->MockedPulsarServiceBaseTest.startBroker:245 ? Runtime
[INFO] 
Error:  org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testMessagePublishBufferThrottleDisabled(org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)
[INFO]   Run 1: PASS
Error:    Run 2: MessagePublishBufferThrottleTest.testMessagePublishBufferThrottleDisabled:50->BrokerTestBase.baseSetup:41->MockedPulsarServiceBaseTest.internalSetup:102->MockedPulsarServiceBaseTest.init:157->MockedPulsarServiceBaseTest.startBroker:245 ? Runtime
[INFO] 
Error:  org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest.testMessagePublishBufferThrottleEnable(org.apache.pulsar.broker.service.MessagePublishBufferThrottleTest)
[INFO]   Run 1: PASS
Error:    Run 2: MessagePublishBufferThrottleTest.testMessagePublishBufferThrottleEnable:80->BrokerTestBase.baseSetup:41->MockedPulsarServiceBaseTest.internalSetup:102->MockedPulsarServiceBaseTest.init:157->MockedPulsarServiceBaseTest.startBroker:245 ? Runtime
[INFO] 
[INFO] 
Error:  Tests run: 3, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  47.758 s
[INFO] Finished at: 2020-11-26T09:01:25Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project pulsar-broker: There are test failures.
Error:  
Error:  Please refer to /__w/pulsar/pulsar/pulsar-broker/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
The command has failed after 3 attempts.
Error: Process completed with exit code 1.
</pre></code>
</detail>

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/MessagePublishBufferThrottleTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/MessagePublishBufferThrottleTest.java
@@ -22,6 +22,7 @@ import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.client.api.Producer;
 import org.apache.pulsar.common.util.FutureUtil;
 import org.testng.Assert;
+import org.testng.annotations.AfterMethod;
 import org.testng.annotations.Test;
 
 import java.util.ArrayList;
@@ -38,9 +39,10 @@ public class MessagePublishBufferThrottleTest extends BrokerTestBase {
         //No-op
     }
 
+    @AfterMethod(alwaysRun = true)
     @Override
     protected void cleanup() throws Exception {
-        //No-op
+        super.internalCleanup();
     }
 
     @Test
@@ -70,7 +72,6 @@ public class MessagePublishBufferThrottleTest extends BrokerTestBase {
         }
         Thread.sleep(20);
         Assert.assertFalse(pulsar.getBrokerService().isReachMessagePublishBufferThreshold());
-        super.internalCleanup();
     }
 
     @Test
@@ -107,7 +108,6 @@ public class MessagePublishBufferThrottleTest extends BrokerTestBase {
             Assert.assertNotNull(future.get());
         }
         Assert.assertEquals(pulsar.getBrokerService().getCurrentMessagePublishBufferSize(), 0L);
-        super.internalCleanup();
     }
 
     @Test
@@ -153,6 +153,5 @@ public class MessagePublishBufferThrottleTest extends BrokerTestBase {
             Assert.assertNotNull(future.get());
         }
         Assert.assertEquals(pulsar.getBrokerService().getCurrentMessagePublishBufferSize(), 0L);
-        super.internalCleanup();
     }
 }

